### PR TITLE
[Fix] mps conv_transpose3d inf-weight NaNs

### LINKED ISSUE #164125: [MPS] `torch.nn.functional.conv_transpose3d` return unexpected results when weight containing `inf`
### 🐛 Describe the bug

Similar to #156062 (cannot reproduce now, may be fixed), `torch.nn.functional.conv_transpose3d` returns all `nan` when the weight contains `inf` on MPS.

```python
import torch
input = torch.zeros((1, 1, 5, 5, 5), dtype=torch.float32)
weight = torch.zeros((1, 1, 3, 3, 3), dtype=torch.float32)
weight[0, 0, 0, 0, 0] = float('inf')

input_mps = input.to("mps")
weight_mps = weight.to("mps")
out_mps = torch.nn.functional.conv_transpose3d(input=input_mps, weight=weight_mps)
out_cpu = torch.nn.functional.conv_transpose3d(input=input, weight=weight)
print(out_cpu) # contains zero
print("--------------------------------")
print(out_mps.cpu()) # all nan
``` 

### Versions

nightly built

### PR DESCRIPTION
**Pull request by SOL1archive**
Fixed #164125

**Update**: Detecting `inf`/`-inf` weights in MPS `conv_transpose3d`, zero them for the main pass, to ensure masks no longer become `NaN`.

I tried my best to fix issue #164164. Since this is my first contribution on PyTorch, I’m open to any feedback or suggestions. Thank you.

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @soulitzer @can-gaa-hou

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 1]
```diff
@@ -634,8 +638,74 @@ Tensor _mps_convolution_transpose(const Tensor& input_t,
...
+  Tensor finite_weight = at::nan_to_num(weight_t, c10::nullopt, 0.0, 0.0);
   auto output_t =
-      mps_convolution_transpose_forward(input_t, weight_t, padding, output_padding, stride, dilation, groups);
+      mps_convolution_transpose_forward(input_t, finite_weight, padding, output_padding, stride, dilation, groups);
+
+  auto pos_inf_mask = weight_t == std::numeric_limits<double>::infinity();
+  auto neg_inf_mask = weight_t == -std::numeric_limits<double>::infinity();
+  const bool has_pos_inf = pos_inf_mask.any().item<bool>();
+  const bool has_neg_inf = neg_inf_mask.any().item<bool>();
+  if (!has_pos_inf && !has_neg_inf) {
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Convolution.mm:654]: Doesn't this check do exactly the same thing as `!inf_mask.any().item<bool>()`, but less efficiently? If so, this check will never be true, as written, since it would have returned earlier.

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 2]
```diff
@@ -634,8 +638,74 @@ Tensor _mps_convolution_transpose(const Tensor& input_t,
       (input_t.dim() == 5 && (input_t.scalar_type() == kHalf || input_t.scalar_type() == kBFloat16));
   TORCH_CHECK(!is_unsupported_3d_dtype, "ConvTranspose 3D with BF16 or FP16 types is not supported on MPS");
 
+  auto inf_mask = at::isinf(weight_t);
+  if (!inf_mask.any().item<bool>()) {
+    return mps_convolution_transpose_forward(input_t, weight_t, padding, output_padding, stride, dilation, groups);
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Convolution.mm:643]: I think the logic seems more complicated than it needs to be. I would suggest only having one call to `mps_convolution_transpose_forward`. I think something along these lines would be more organized:

```
bool mask_has_inf = ...;
auto finite_weight = (mask_has_inf) ? ... : weight_t;
auto output_t = mps_convolution_transpose_forward(...);
if (mask_has_inf) {
    // any additional things that need to be done here
    ...
}
return output_t;
```

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 3]
```diff
@@ -634,8 +638,74 @@ Tensor _mps_convolution_transpose(const Tensor& input_t,
...
+    if (grad_neg_any) {
+      minus_inf_mask = minus_inf_mask | run_mask_conv(grad_neg_mask, pos_inf_mask);
+    }
+  }
+  if (has_neg_inf) {
+    if (grad_pos_any) {
+      minus_inf_mask = minus_inf_mask | run_mask_conv(grad_pos_mask, neg_inf_mask);
+    }
+    if (grad_neg_any) {
+      plus_inf_mask = plus_inf_mask | run_mask_conv(grad_neg_mask, neg_inf_mask);
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Convolution.mm:690]: It looks like in the worst case, if `weights_t` has both positive and negative inf values and `input_t` has both negative and positive values, then we'll end up calling `mps_convolution_transpose_forward` four additional times. Is that really necessary? It's not totally clear to me why all this extra work is being done

File: test/test_mps.py [hunk 4]
```diff
@@ -828,6 +828,37 @@ def test_conv_high_channel_size(self):
...
+        mixed_input = torch.zeros((1, 2, 1, 1, 1), dtype=torch.float32)
+        mixed_input[0, 0, 0, 0, 0] = 1.0
+        mixed_input[0, 1, 0, 0, 0] = 1.0
+        mixed_weight = torch.zeros((2, 1, 1, 1, 1), dtype=torch.float32)
+        mixed_weight[0, 0, 0, 0, 0] = float("inf")
+        mixed_weight[1, 0, 0, 0, 0] = float("-inf")
+
+        cpu_nan = F.conv_transpose3d(mixed_input, mixed_weight)
+        mps_nan = F.conv_transpose3d(mixed_input.to("mps"), mixed_weight.to("mps")).cpu()
+        self.assertEqual(cpu_nan.isnan(), mps_nan.isnan())
```
Comment (kurtamohler) [inline test/test_mps.py:860]: I think we actually want `assert_match(mixed_input, mixed_weight)`, right? If the whole thing doesn't match, then the issue isn't solved.

Also, I think we should test several more cases where the weight has infs. We should try several differently sized sets of inputs, perhaps with `@parametrize`. And maybe we should randomize which elements get set to inf as well.

File: aten/src/ATen/native/mps/operations/Convolution.mm (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/operations/Convolution.mm:643]: But actually, wouldn't it be better to make `mps_convolution_backward_input` handle infs correctly, so that all of its callers benefit from the change?

**Review by kurtamohler — COMMENTED** [review]
Thanks for the PR!

I'm concerned that this could significantly impact performance. It would be good to check that with `torch.utils.benchmark`, with several different sets of sizes for the inputs.

One of the reasons this may impact performance is because of the fact that the host is checking values of MPS tensors, which requires a sync. I think it's probably possible to do all the work in MPS Graph instead, avoiding a sync
